include/dr/shp/algorithms/for_each.hpp

```diff
@@ -16,74 +16,23 @@ namespace shp {
 template <typename ExecutionPolicy, lib::distributed_contiguous_range R,
           typename Fn>
 void for_each(ExecutionPolicy &&policy, R &&r, Fn &&fn) {
-
-  namespace sycl = cl::sycl;
-
-  static_assert(
+  static_assert( // currently only one policy supported
       std::is_same_v<std::remove_cvref_t<ExecutionPolicy>, device_policy>);
 
-  if constexpr (std::is_same_v<std::remove_cvref_t<ExecutionPolicy>,
-                               device_policy>) {
-    auto &&devices = std::forward<ExecutionPolicy>(policy).get_devices();
-
-    std::vector<sycl::queue> queues;
-    std::vector<sycl::event> events;
-
-    for (auto &&segment : lib::ranges::segments(r)) {
-      auto device = devices[lib::ranges::rank(segment)];
-
-      sycl::queue q(device);
+  auto devices = policy.get_devices();
+  std::vector<cl::sycl::event> events;
 
-      auto begin = lib::ranges::local(rng::begin(segment));
-
-      assert(rng::size(segment) > 0);
-      auto event = q.parallel_for(rng::size(segment),
-                                  [=](sycl::id<1> idx) { fn(*(begin + idx)); });
-      events.emplace_back(event);
-      queues.emplace_back(q);
-    }
-
-    for (auto &&event : events) {
-      event.wait();
-    }
-  } else {
-    assert(false);
+  for (auto &&segment : lib::ranges::segments(r)) {
+    assert(rng::size(segment) > 0);
+    events.emplace_back(
+        sycl::queue(devices[lib::ranges::rank(segment)])
+            .parallel_for(rng::size(segment), [=](cl::sycl::id<1> idx) {
+              fn(*(rng::begin(segment) + idx));
```

Comment:
It's actually important that we extract the iterator and only copy that to the GPU.  The segment itself is likely not device copyable.  It's also important that we call `local()` on the iterator to get a raw pointer, as there's a bug in our SYCL runtime that may keep this from running otherwise.